allennlp/training/learning_rate_schedulers.py

```diff
@@ -126,6 +126,99 @@ def get_lr(self):
+    """
+    Implements the Slanted Triangular Learning Rate schedule with optional gradual
+    unfreezing. The schedule corresponds to first linearly increasing the learning
+    rate and annealing the learning based on a fixed ratio.
+    If we gradually unfreeze, then in the first epoch of training, only the top
+    layer is trained; in the second epoch, the top two layers are trained, etc.
+    During freezing, the learning rate is increased and annealed over one epoch.
+    After freezing finished, the learning rate is increased and annealed over
+    the remaining training iterations.
+    Note that with this schedule, early stopping should typically be avoided.
+
+    Parameters
+    ----------
+    num_epochs : ``int``, required.
+        The total number of epochs for which the model should be trained.
+    num_steps_per_epoch: ``int``, required.
+        The number of steps (updates, batches) per training epoch.
+    cut_frac: ``float``, optional (default = 0.1).
+        The fraction of the steps to increase the learning rate.
+    ratio: ``float``, optional (default = 32).
+        The ratio of the smallest to the (largest) base learning rate.
+    gradual_unfreezing: ``bool``, optional (default = False).
+        Whether gradual unfreezing should be used.
+    """
+    def __init__(self,
+                 optimizer: torch.optim.Optimizer,
+                 num_epochs: int,
+                 num_steps_per_epoch: int,
+                 cut_frac: float = 0.1,
+                 ratio: int = 32,
+                 lr: float = 0.01,
+                 last_epoch: int = -1,
+                 gradual_unfreezing: bool = False) -> None:
+        self.num_epochs = num_epochs
+        self.num_steps_per_epoch = num_steps_per_epoch
+        self.cut_frac = cut_frac
+        self.ratio = ratio
+        self.rate = lr
+        self.gradual_unfreezing = gradual_unfreezing
+        self.freezing_current = self.gradual_unfreezing
+        self.first_epoch = True
+        super().__init__(optimizer, last_epoch=last_epoch)
+
+    def step(self, epoch=None):
+        if self.gradual_unfreezing:
+            assert not self.optimizer.param_groups[-1]["params"], \
+                "The default group should be empty."
+            # the method is called once when initialising before the
+            # first epoch (epoch 0) and then always at the end of each
+            # epoch; so the first time, with epoch id 0, we want to set
+            # up for epoch #1; the second time, still with epoch id 0,
+            # we want to set up for epoch #2, etc.
+            num_layers_to_unfreeze = epoch+1 if self.first_epoch else epoch+2
+            if self.first_epoch:
+                self.step_batch(1)
+                self.first_epoch = False
+            if num_layers_to_unfreeze >= len(self.optimizer.param_groups)-1:
+                print('Gradual unfreezing finished. Training all layers.')
```

Comment:
it'd be good if these print statements were `logger.info` calls instead.

Reply:
Done.